packages/react/src/use-chat.ts

```diff
@@ -522,18 +525,63 @@ By default, it's set to 0, which will disable the feature.
 
       event?.preventDefault?.();
 
+      const attachmentsForRequest: Attachment[] = [];
+      const attachmentsFromOptions = options.experimental_attachments;
+
+      if (attachmentsFromOptions) {
+        if (attachmentsFromOptions instanceof FileList) {
+          for (const attachment of Array.from(attachmentsFromOptions)) {
+            const { name, type } = attachment;
+
+            const dataUrl = await new Promise<string>((resolve, reject) => {
+              const reader = new FileReader();
+              reader.onload = readerEvent => {
+                resolve(readerEvent.target?.result as string);
+              };
+              reader.onerror = error => reject(error);
+              reader.readAsDataURL(attachment);
+            });
+
+            attachmentsForRequest.push({
+              name,
+              mimeType: type,
+              url: dataUrl,
+            });
+          }
+        } else if (Array.isArray(attachmentsFromOptions)) {
+          for (const file of attachmentsFromOptions) {
+            const { name, url, mimeType } = file;
+
+            attachmentsForRequest.push({
+              name,
+              mimeType,
+              url,
+            });
+          }
+        } else {
+          throw new Error('Invalid attachments type');
+        }
+      }
+
       const requestOptions = {
         headers: options.headers ?? options.options?.headers,
         body: options.body ?? options.options?.body,
       };
 
       const chatRequest: ChatRequest = {
         messages: input
-          ? messagesRef.current.concat({
-              id: generateId(),
-              role: 'user',
-              content: input,
-            })
+          ? options.experimental_attachments
+            ? messagesRef.current.concat({
+                id: generateId(),
+                role: 'user',
+                content: input,
+                experimental_attachments: attachmentsForRequest,
+              })
+            : messagesRef.current.concat({
+                id: generateId(),
+                role: 'user',
+                content: input,
+              })
```

Comment:
wondering if we can simplify this.
```
messagesRef.current.concat({
  id: generateId(),
  role: 'user',
  content: input,
  experimental_attachments: attachmentsForRequest.length > 0 ? attachmentsForRequest : undefined
})
```

Reply:
reworked this part of code to pass `experimental_attachments` and filter it out when constructing the request payload [here](https://github.com/vercel/ai/pull/2226/files#diff-0a57b9a56351ecd277e650d5a02072fbb65c6d685b81103e1cdf0459cf7c323eR123-R125)